### PR TITLE
Recover Tokens

### PR DESCRIPTION
**Pull request by AStox**
This PR adds functionality for governance to recover tokens that were accidentally sent to the wrong contract

### CODE REVIEW

File: src/libraries/MessagesLib.sol [hunk 1]
```diff
@@ -504,6 +506,18 @@ library MessagesLib {
         return (_msg.toBytes32(1));
     }
 
+    /**
+    * Recover Tokens sent to the wrong contract
+    *
+    * 0: call type (uint8 = 1 byte)
+    * 1-20: The contract address (address = 20 bytes)
+    * 21-40: The recipient address (address = 20 bytes)
```
Comment (hieronx) [inline src/libraries/MessagesLib.sol:514]: These should be 32 bytes, of which the first 20 bytes are parsed as an address, for compatibility with future non EVM use cases

File: src/LiquidityPool.sol [hunk 2]
```diff
@@ -90,6 +90,10 @@ contract LiquidityPool is Auth, IERC7540 {
         emit File(what, data);
     }
 
+    function recoverTokens(address to, uint256 amount) external auth {
```
Comment (hieronx) [inline src/LiquidityPool.sol:93]: We should add asset as another param. Both for non liquidity pool contracts we add it to, but even for liquidity pool, someone can accidentally send USDC to the FRAX liquidity pool

**Review by hieronx — COMMENTED** [review]
Nice work!

File: src/LiquidityPool.sol (hunk 2 above)
Comment (AStox) [reply src/LiquidityPool.sol:93]: Great point. I'll update it

File: src/LiquidityPool.sol [hunk 3]
```diff
@@ -90,6 +90,10 @@ contract LiquidityPool is Auth, IERC7540 {
         emit File(what, data);
     }
 
+    function recoverTokens(address to, address token, uint256 amount) external auth {
```
Comment (hieronx) [inline src/LiquidityPool.sol:93]: Would be cleaner if the order of the params of the method matches the order of the transfer func
```suggestion
    function recoverTokens(address token, address to, uint256 amount) external auth {
```

File: src/Root.sol [hunk 4]
```diff
@@ -30,6 +34,7 @@ contract Root is Auth {
     event CancelRely(address indexed target);
     event RelyContract(address indexed target, address indexed user);
     event DenyContract(address indexed target, address indexed user);
+    event RecoverTokens(address indexed fromContract, address indexed toUser, address indexed token, uint256 amount);
```
Comment (hieronx) [inline src/Root.sol:37]: Compared to other messages, plus using the transfer method ordering, I would name it like this now:
```suggestion
    event RecoverTokens(address indexed target, address indexed token, address indexed to, uint256 amount);
```

File: test/integration/Admin.t.sol [hunk 5]
```diff
@@ -260,4 +260,22 @@ contract AdminTest is BaseTest {
         root.relyContract(address(investmentManager), address(this));
         assertEq(investmentManager.wards(address(this)), 1);
     }
+
+    //------ Token Recovery tests ------///
+    function testRecoverTokensFromLiquidityPool() public {
+        deploySimplePool();
+        address clumsyUser = vm.addr(0x1234);
```
Comment (hieronx) [inline test/integration/Admin.t.sol:267]: 😄

File: src/gateway/Gateway.sol [hunk 6]
```diff
@@ -90,6 +90,7 @@ interface RootLike {
     function paused() external returns (bool);
     function scheduleRely(address target) external;
     function cancelRely(address target) external;
+    function recoverTokens(address fromContract, address toUser, address token, uint256 amount) external;
```
Comment (hieronx) [inline src/gateway/Gateway.sol:93]:
```suggestion
    function recoverTokens(address target, address token, address to, uint256 amount) external;
```

File: test/mocks/MockCentrifugeChain.sol [hunk 7]
```diff
@@ -159,6 +159,17 @@ contract MockCentrifugeChain is Test {
         _execute(_message);
     }
 
+    function recoverTokens(address fromContract, address toUser, address token, uint256 amount) public {
+        bytes memory _message = abi.encodePacked(
+            uint8(MessagesLib.Call.RecoverTokens),
+            bytes32(bytes20(fromContract)),
+            bytes32(bytes20(toUser)),
+            bytes32(bytes20(token)),
```
Comment (hieronx) [inline test/mocks/MockCentrifugeChain.sol:167]: There's a helper for this in `BytesLib`:
```suggestion
            fromContract.toBytes32(),
            toUser.toBytes32(),
            token.toBytes32(),
```

File: test/mocks/MockCentrifugeChain.sol [hunk 8]
```diff
@@ -159,6 +159,17 @@ contract MockCentrifugeChain is Test {
         _execute(_message);
     }
 
+    function recoverTokens(address fromContract, address toUser, address token, uint256 amount) public {
```
Comment (hieronx) [inline test/mocks/MockCentrifugeChain.sol:162]: You haven't swapped the param names here, but are actually calling this with token before toUser  already from the test
```suggestion
    function recoverTokens(address target, address token, address to, uint256 amount) public {
```

**Review by hieronx — APPROVED** [review]
Looks good!

**Review by hieronx — DISMISSED** [review]

File: src/gateway/Gateway.sol [hunk 9]
```diff
@@ -93,6 +94,8 @@ contract Gateway is Auth {
             manager = poolManager;
         } else if (id == 27) {
             manager = investmentManager;
+        } else if (id == 31) {
```
Comment (ilinzweilin) [inline src/gateway/Gateway.sol:97]: I would add it as an "or" statement and combine with the other root cases

**Review by ilinzweilin — APPROVED** [review]

**Review by ilinzweilin — APPROVED** [review]

**Pull request merged**